glommio/src/sys/uring.rs

```diff
@@ -454,32 +455,30 @@ fn transmute_error(res: io::Result<u32>) -> io::Result<usize> {
         })
 }
 
-fn process_one_event<F>(
-    cqe: Option<iou::CQE>,
-    try_process: F,
-    wakers: &mut Vec<Waker>,
-) -> Option<()>
+fn process_one_event<F>(cqe: Option<iou::CQE>, try_process: F) -> Option<bool>
 where
     F: FnOnce(&InnerSource) -> Option<()>,
 {
     if let Some(value) = cqe {
         // No user data is POLL_REMOVE or CANCEL, we won't process.
         if value.user_data() == 0 {
-            return Some(());
+            return Some(false);
         }
 
         let src = consume_source(from_user_data(value.user_data()));
 
         let result = value.result();
 
+        let mut woke = false;
         if try_process(&*src).is_none() {
```

Comment:
I think this needs work as well.
`try_process()` essentially gives the opportunity for the ring to process events that are internal to the ring architecture and therefore may be handled differently (one general comment about that: as uring.rs evolved, there is probably a way now to remove this special case!)

However if those events fire, I suspect we may also have to forego sleeping.

This code is a bit complex, so let's be careful: due to cancellations, it *may* be that `LinkRings` actually doesn't have to prevent waking. I have to inspect that code a bit more carefully again.

However, even if that's the case, I'd recommend that `try_process` returns a boolean in this case, and `LinkRings` return false.

Reply:
From my understanding, `try_process()` is for the ring to decide whether an event should be waking or not. And in the current code, both events `LinkRings` and `Timeout` have no waker inside, which means removing the `try_process()` parameter will not change the ring's behavior. I tried it and the test cases pass (several times).

Reply:
> However if those events fire, I suspect we may also have to forego sleeping.

We always sleep on the main ring, therefore I think it is OK to not consider this situation as we won't see the `LinkRings` event of this time "going to sleep"?

Reply:
I think we should not indeed increment the waker on `LinkRings`. My main concern is with the interface.

However, let's leave this as is: I'll actually try to get rid of this special case altogether.
With the timer gone in the PR I linked, it's only one more to go...